kasa/tapo/tapodevice.py

```diff
@@ -0,0 +1,157 @@
+"""Module for a TAPO device."""
+import base64
+import logging
+from datetime import datetime, timedelta, timezone
+from typing import Any, Dict, Optional, Set
+
+from ..aesprotocol import TPLinkAes
+from ..exceptions import AuthenticationException
+from ..smartdevice import SmartDevice
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class AttributeDict(dict):
+    """Class to allow a dict to be accessed as if keys are variables."""
+
+    __getattr__ = dict.__getitem__
+    __setattr__ = dict.__setitem__  # type: ignore[assignment]
+    __delattr__ = dict.__delitem__  # type: ignore[assignment]
+
+
+# mypy: disable-error-code="attr-defined"
+class TapoDevice(SmartDevice):
+    """Base class to represent a TAPO device."""
+
+    def __init__(
+        self,
+        host: str,
+        *,
+        port: Optional[int] = None,
+        credentials,
+        timeout: Optional[int] = None,
+    ) -> None:
+        super().__init__(host, port=port, credentials=credentials, timeout=timeout)
+        self._tapo_client = None
+        self._state_information: Dict[str, Any] = {}
+        self.protocol = TPLinkAes(host, credentials=credentials, timeout=timeout)
+
+    async def update(self, update_children: bool = True):
+        """Update the device."""
+        if self.credentials is None or self.credentials.username is None:
+            raise AuthenticationException("Tapo plug requires authentication.")
+
+        self._info = AttributeDict(await self.protocol.query("get_device_info"))
+        self._usage = AttributeDict(await self.protocol.query("get_device_usage"))
+        self._time = AttributeDict(await self.protocol.query("get_device_time"))
+
+        self._last_update = self._data = {
+            "info": self._info,
+            "usage": self._usage,
+            "time": self._time,
+        }
+
+        self._update_state_information()
+
+        _LOGGER.debug("Got an update: %s", self._data)
+
+    def _update_state_information(self):
+        self._state_information = {
+            "is_hw_v2": self._info.hw_ver == "2.0",
+            "overheated": self._info.overheated,
+            "signal_level": self._info.signal_level,
+            "auto_off_status": self._info.auto_off_status,
+            "auto_off_remain_time": self._info.auto_off_remain_time,
```

Comment:
These two settings are not available on my bulb, so probably a good idea to move them out from the base class, and just overload and call the parent information inside the plug class.

I think we should start thinking a solution for the exposing extra features like these to homeassistant and other downstreams using a common descriptor-type interface, but that's not something for this PR.

Reply:
Done